### ISSUE TITLE
What does a replay look like?

### ISSUE THREAD

**Issue by mattzeunert** [nodejs/node-chakracore#157]
I'm recording the execution of a script like this:

    ./node -TTRecord:log test.js 

Then I replay it:

    ./node -TTReplay:log
    > 
    Reached end of Execution -- Exiting.

What is supposed to happen when running a replay?

If I try to use VS Code instead this is the output:

    node --debug-brk=9920 --nolazy -TTDebug:log -TTBreakFirst test.js 
    Debugger listening on 127.0.0.1:9920
    Reached end of Execution -- Exiting. 

It seems to momentarily pause in `events.js` and then quit.

![time-travel](https://cloud.githubusercontent.com/assets/1303660/21289284/d2d7faea-c490-11e6-8742-98ffabe04834.gif)

This is on Mac OS. `test.js` is a simple one-off script, not a server like in the RFSDemo code, but I assume that isn't important?

**Comment by mrkmarron** [comment]
Hi, from you description my hypothesis is that what you are seeing is a design choice we made with (1) when the trace recording starts and/or (2) our choice to suppress console log output in replay. 

Since startup code is easy to re-run and generally involves a lot of (uninteresting) initialization work we do *not* start recording until this code completes and the runtime moves into the event loop callbacks. From your gif it looks all the breakpoints are set in the main function (which I assume is executed as part of initialization) so they won't be hit in the replay. However, there is some callback based execution (in the event emitter) which does execute in replay and is the momentary pause. If you set a breakpoint in this code you should see the debugger break as expected.

For console log output we have a special "telemetryLog" function with the signature telemetryLog(msgString, printFlag). To have it print during replay as well as record the "printFlag" variable can be set to true.

If this doesn't clear things up please let me know and if you share the code sample I am happy to take a look at it personally.

Mark

**Comment by mattzeunert** [comment]
Thanks for your answer, Mark! It's awesome to see time-travel debugging becoming viable.

I'm able to pause somewhere now when using VS Code, but I'm still not sure how to use it.

I've got this code in simple.js:

```
setTimeout(function(){
    var a = 5;
    a += 6;
    debugger;
    telemetryLog(a.toString(), true);
}, 200)
```
Then I run:

    ./node -TTRecord:log simple.js // prints out 11
    ./node -TTReplay:log simple.js  // also prints out 11
    ./node -TTReplay:log -TTBreakFirst simple.js 
    
Which gives me this error:

```
TypeError: Object doesn't support property or method 'setBreakPoint'
   at Module.prototype._compile (module.js:566:7)
   at Module._extensions[.js] (module.js:583:3)
   at Module.prototype.load (module.js:491:3)
   at tryModuleLoad (module.js:450:5)
   at Module._load (module.js:442:3)
   at Module.runMain (module.js:608:3)
   at run (bootstrap_node.js:382:7)
   at startup (bootstrap_node.js:137:9)
   at Anonymous function (bootstrap_node.js:497:3)
```

I don't have any expectations about what should happen, but it's not clear from the [TTD readme](https://github.com/nodejs/node-chakracore/blob/xplat/TTD-README.md) how to use the replay.

When I try to use VS Code it pauses on the debugger statement, but if I try to use any of the stepping controls I get an error:

    node --debug-brk=28555 -TTDebug:log -TTBreakFirst simple.js 
    Debugger listening on 127.0.0.1:28555
    TTD assert failed: Push/Pop is not matched so something went wrong.

**Comment by mrkmarron** [comment]
The push/pop assert is definitely a bug with how we interact with the debugger in a specific case (we are missing some logic in an exception flow path). I will get a fix checked in shortly and also see if I can point out the exact feature in your example that is triggering it so you can workaround until the fix lands.

The replay with --TTBreakFirst is due to us not allowing debugger attach in replay mode (we should have a more informative error though). Replay mode always re-executes the recorded events/code and then exits so most users will want to use the -TTDebug for debugging but there are cases where being able to replay the execution and gather additional information about it is useful. For example you could replay the code on an instrumented version of node-chakracore that records how many times each loop is executed, which loops are jitted, etc. 

Thanks for trying things out and reporting these issues.

**Comment by mrkmarron** [comment]
A fix for the "TTD assert failed: Push/Pop is not matched so something went wrong." was merged in PR #158 and I have verified that everything works in the timeout example (including use of debugger statement). I also opened an issue around the usage of the -TTBreakFirst flag as issue #159 for easier tracking.

**Comment by mattzeunert** [comment]
That's awesome! Working really well in VS Code now.
